tests/modules/blockedclient.c

```diff
@@ -155,6 +153,9 @@ void *bg_call_worker(void *arg) {
     RedisModule_Free(bg->argv);
     RedisModule_Free(bg);
 
```

Comment:
start version: 6.2.0
introduced: #8141
reason: release client argv outside of GIL

```
WARNING: ThreadSanitizer: data race (pid=1969717)
  Write of size 4 at 0x558d01163de0 by thread T14 (mutexes: write M64):
    #0 processEventsWhileBlocked /data/redis_fork/src/networking.c:4116 (redis-server+0xe795d)
    #1 RM_Yield /data/redis_fork/src/module.c:2410 (redis-server+0x20f4c3)
    #2 RM_Yield /data/redis_fork/src/module.c:2381 (redis-server+0x20f4c3)
    #3 bg_call_worker /data/redis_fork/tests/modules/blockedclient.c:116 (blockedclient.so+0x9ade)

  Previous read of size 4 at 0x558d01163de0 by main thread (mutexes: write M86, write M82, write M84):
    #0 afterSleep /data/redis_fork/src/server.c:1867 (redis-server+0x9f4ab)
    #1 aeProcessEvents /data/redis_fork/src/ae.c:408 (redis-server+0x971dc)
    #2 aeMain /data/redis_fork/src/ae.c:496 (redis-server+0x971dc)
    #3 main /data/redis_fork/src/server.c:7212 (redis-server+0x846d5)
```

Reply:
I update the stack trace.

Reply:
so what do we conclude here?
that if a module retains the argv strings, it must release them before unblocking the client, or alternatively with the GIL locked?
@MeirShpilraien PTAL

Reply:
actually the order doesn't matter, if they both decr at the same time it can mess up the refcount.

Reply:
i suppose we better document this.
i.e. that anything we do with the retained arguments refcount) from a thread (using them for a reply, re-retaining, or freeing them) needs to be done with the GIL locked.
@MeirShpilraien @yossigo WDYT?

Reply:
@oranagra Technically we can document that, but does that make sense in terms of API usability?

Reply:
what other options do we have? introduce a new API that deep copies the string to be used when retaining it for a thread? use atomic operations on all refcounts?
the only relief here is that modules that use threads to process argv are probably not too common.

Reply:
I think its OK to say that release a RedisModuleString can only be done when holding the GIL unless you know for sure that you are the only owner of the RedisModuleString.

Reply:
i see we updated the tests, but we also need to document that.
question is where? should that be done in RM_RetainString, RM_BlockClient, RM_ThreadSafeContextLock?

@MeirShpilraien do you have a better idea?

Reply:
I think we should mention it on `RM_FreeString`, `RM_RetainString`, and `RM_HoldString`. We should mentioned that those function are not thread safe and should only be called when the GIL is held. We should decide if we want to document the exception about the `RM_FreeString` in case you know you are the only owner.

Reply:
@sundb this one is linked to number 4 at the top, which says:
> Harm Level: None
Trigger assertion

first of all, an assertion isn't "None" it terminates the process and can cause data loss. maybe change it to "Low"?
secondly this can also mess up the refcount, or cause double free and other issues that can lead to memory corruption.
am i missing anything?

Reply:
my intention was that it was only caused by the module's use of API, not by the internal implement, so I chose `None`.
let's change it to `LOW`.

Reply:
ohh, ok, maybe in addition to the "Harm Level", we can add some explicit "Trigger", to specify that it depends on some rare case, and people can easily rule it out and know they're safe.
or we can leave it as is, the scenario is there, it's just a little bit hard to understand if you're not keen on the details.